### PR TITLE
[ICDS Dashboard] Added vhnds api for governance dashboard

### PR DESCRIPTION
**Pull request by challabeehyv**
https://app.asana.com/0/0/1154180297098374/f

##### SUMMARY
Added vhnds api for governance dashboard

##### FEATURE FLAG
No

### CODE REVIEW

File: custom/icds_reports/models/aggregate.py [hunk 1]
```diff
@@ -1550,3 +1551,21 @@ class Meta(object):
...
+
+
+class AggregateVHNDForms(models.Model, AggregateMixin):
+    state_id = models.TextField(null=True)
+    vhsnd_date_past_month = models.DateField(null=True)
+    anm_mpw_present = models.NullBooleanField(null=True)
+    asha_present = models.NullBooleanField(null=True)
+    child_immu = models.NullBooleanField(null=True)
+    anc_today = models.NullBooleanField(null=True)
+    awc_id = models.TextField(null=True)
```
Comment (Rohit25negi) [inline custom/icds_reports/models/aggregate.py]: would be good to mark Awc_id  as primary key else django will add primary key column by itself

File: custom/icds_reports/reports/governance_apis.py [hunk 2]
```diff
@@ -41,6 +52,40 @@ def base_data(row_data):
...
+    ).order_by(*order).values(
+        'state_name', 'district_name', 'block_name', 'supervisor_name', 'awc_name', 'month',
+        'vhsnd_date_past_month', 'anm_mpw_present', 'asha_present', 'child_immu', 'anc_today', 'awc_id'
+    )
+
+    paginated_data = data[:length]
+
+    def base_data(row_data):
+        return dict(
+            state=get_value_or_data_not_entered(row_data, 'state_name'),
```
Comment (Rohit25negi) [inline custom/icds_reports/reports/governance_apis.py]: I think this can be simplified like this https://github.com/dimagi/commcare-hq/pull/26325/files#diff-2c73ea571072adbc6bb48114d0b66e7dR48. It will be much less code and more readable

File: custom/icds_reports/models/aggregate.py [hunk 3]
```diff
@@ -1550,3 +1551,21 @@ class Meta(object):
...
+    month = models.DateField(null=True)
+
+    class Meta(object):
+        db_table = AGG_GOV_VHND_TABLE
+        unique_together = ('month', 'awc_id')
```
Comment (Rohit25negi) [inline custom/icds_reports/models/aggregate.py]: Good to Add State in Unique together as the filter will also include state_id. it will make the query to use indexes

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        return """
+        INSERT INTO "{tablename}" (
+          vhsnd_date_past_month, state_id, anm_mpw_present, asha_present, child_immu, anc_today,
+          awc_id, month
+        )(
+        SELECT
+        DISTINCT ON(awc_id, month) vhsnd_date_past_month, state_id,
+         CASE WHEN(anm_mpw = 1) THEN TRUE  ELSE FALSE END as anm_mpw_present,
```
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: Could not this be replace by `anm_mpw = 1 as anm_mpw_present`?

File: custom/icds_reports/tests/agg_tests/test_governance_apis.py [hunk 5]
```diff
@@ -83,3 +83,60 @@ def test_data_fetching_state_names_api(self):
...
+        data, count = get_vhnd_data(limit, 2017, 5, order, query_filters)
+        expected_first_row = {
+            "state": "st1", "district": "d1", "block": "b1", "sector": "s1", "awc": "a49",
+            "month": datetime.date(2017, 5, 1), "vhsnd_conducted": "no", "vhsnd_date": "Data Not Entered",
+            "anm_present": "no", "asha_present": "no", "any_child_immunized": "no",
+            "anc_conducted": "no"
+        }
+        self.assertEqual(data[0], expected_first_row)
+
+    def test_data_fetching_total_record_count_for_no_records_for_vhnds_api(self):
```
Comment (Rohit25negi) [inline custom/icds_reports/tests/agg_tests/test_governance_apis.py]: Can you also add a test case to test if there are multiple vhnds in a moth for one awc it is giving you the first one. That will also be be the validation of your aggregation code.

File: custom/icds_reports/tasks.py [hunk 6]
```diff
@@ -663,6 +664,12 @@ def _agg_ls_vhnd_form(state_id, day):
         AggregateLsVhndForm.aggregate(state_id, force_to_date(day))
 
 
+@track_time
+def _agg_gov_vhnd_form(state_id, day):
```
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: Could you also add this function to the pipe repository? Here is the example Pr: https://github.com/dimagi/pipes/pull/24.
you would have to first add this function in `custom/icds_reports/management/commands/run_aggregation_query.py`

File: custom/icds_reports/reports/governance_apis.py [hunk 7]
```diff
@@ -41,6 +52,40 @@ def base_data(row_data):
     return data_rows, data.count()
 
 
+@icds_quickcache(['length', 'year', 'month', 'order', 'query_filters'], timeout=30 * 60)
+def get_vhnd_data(length, year, month, order, query_filters):
+    data = GovVHNDView.objects.filter(
+        month=date(year, month, 1),
+        **query_filters
+    ).order_by(*order).values(
```
Comment (Rohit25negi) [inline custom/icds_reports/reports/governance_apis.py]: Can you update the values returning based on the revised spec.

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
@challabeehyv  Left  number of change requests.

File: custom/icds_reports/reports/governance_apis.py (hunk 2 above)
Comment (challabeehyv) [reply custom/icds_reports/reports/governance_apis.py]: there are some changes to the return names and value formats. So added this way

File: custom/icds_reports/models/aggregate.py (hunk 1 above)
Comment (challabeehyv) [reply custom/icds_reports/models/aggregate.py]: done

File: custom/icds_reports/models/aggregate.py (hunk 3 above)
Comment (challabeehyv) [reply custom/icds_reports/models/aggregate.py]: added

File: custom/icds_reports/tests/agg_tests/test_governance_apis.py (hunk 5 above)
Comment (challabeehyv) [reply custom/icds_reports/tests/agg_tests/test_governance_apis.py]: added

File: custom/icds_reports/reports/governance_apis.py (hunk 7 above)
Comment (challabeehyv) [reply custom/icds_reports/reports/governance_apis.py]: added

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 4 above)
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: yeah.makes sense

File: custom/icds_reports/reports/governance_apis.py (hunk 2 above)
Comment (Rohit25negi) [reply custom/icds_reports/reports/governance_apis.py]: YOu could use the F function as in the above link

File: custom/icds_reports/tasks.py (hunk 6 above)
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: done

File: custom/icds_reports/reports/governance_apis.py (hunk 2 above)
Comment (challabeehyv) [reply custom/icds_reports/reports/governance_apis.py]: done

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+          month
+        )(
+        SELECT
+          DISTINCT awc_id as awc_id,
+          FIRST_VALUE(vhsnd_date_past_month) over w as vhsnd_date_past_month,
+          FIRST_VALUE(state_id) over w as state_id,
+          FIRST_VALUE(anm_mpw=1) over w as anm_mpw_present,
+          FIRST_VALUE(asha_present=1) over w as asha_present,
+          FIRST_VALUE(child_immu=1) over w as child_immu,
+          FIRST_VALUE(anc_today=1) over w as anc_today,
```
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: not sure. But does `FIRST_VALUE(anc_today=1)` give the same result as `FIRST_VALUE(anc_today)=1`?

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+          %(start_date)s::DATE AS month
+          FROM "{ucr_tablename}"
+            WHERE vhsnd_date_past_month >= %(start_date)s AND
+            vhsnd_date_past_month < %(end_date)s AND state_id = %(state_id)s
+          WINDOW w AS (
+                PARTITION BY awc_id
+                ORDER BY submitted_on
```
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: it should be `order by vhsnd_date_past_month RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING`

File: custom/icds_reports/views.py [hunk 10]
```diff
@@ -2283,3 +2284,30 @@ class GovernanceStateListAPI(GovernanceAPIBaseView):
...
+        is_valid, error_message = self.validate_param(state_id, month, year)
+
+        if not is_valid:
+            return HttpResponse(error_message, status=400)
+
+        query_filters = {'awc_id__gt': last_awc_id}
+        order = ['awc_id']
+        if state_id != '':
+            query_filters['state_id'] = state_id
+        data, count = get_vhnd_data(GOVERNANCE_API_HOME_VISIT_RECORDS_PAGINATION, year,
```
Comment (Rohit25negi) [inline custom/icds_reports/views.py]: could you change the name of this constant to be more generic instead of just home visit.

File: custom/icds_reports/tasks.py [hunk 11]
```diff
@@ -1741,6 +1753,16 @@ def update_dashboard_activity_report(target_date=None):
         DashboardUserActivityReport().aggregate(target_date)
 
 
+def update_vhnd_form_aggregation(state_id, target_month=None):
```
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: Why this function is created?

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
+from dateutil.relativedelta import relativedelta
+
+from custom.icds_reports.const import AGG_GOV_VHND_TABLE
+from custom.icds_reports.utils.aggregation_helpers import month_formatter
+from custom.icds_reports.utils.aggregation_helpers.distributed.base import (
+    StateBasedAggregationPartitionedHelper,
+)
+
+
+class GovVhndFormAggDistributedHelper(StateBasedAggregationPartitionedHelper):
```
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: I think despite of me saying to create a separate table and aggregate it. I think it may make sense to just add this aggregation as a part of `AggGovDashboardHelper` and introduce all-new columns in that table. We will gain the following benefits from it.
1) Since now are already aggregating gov dashboard separately, it does not make sense to let this slow down the standard aggregation when we can make this part of gov dashboard aggregation.
2) All gov dashboard related info will be in a single table which will be isolated from standard dashboard.

Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: It will also reduce your code change size. And size of the Gov_dashboard table is significantly smaller than agg_awc so its faster to query from that.

Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: But my main moto is not to slow down the standard aggregation whenever possible.

File: custom/icds_reports/tasks.py [hunk 13]
```diff
@@ -664,6 +670,12 @@ def _agg_ls_vhnd_form(state_id, day):
         AggregateLsVhndForm.aggregate(state_id, force_to_date(day))
 
 
+@track_time
+def _agg_gov_vhnd_form(state_id, day):
```
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: After shifting the aggregation to gov dashboard table. you wont need this function.

File: custom/icds_reports/const.py [hunk 14]
```diff
@@ -165,5 +167,6 @@ class MapColors(object):
 # need to add more steps
 GOVERNANCE_API_STEPS = [
```
Comment (Rohit25negi) [inline custom/icds_reports/const.py]: do you still need this constant?

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
codewise looks good.
Left a few structural comments.

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 8 above)
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: this gives syntax error. First value will take expression/value as input like FIRST_VALUE(expression). The expression can be an expression, column, or subquery evaluated against the value of the first row of a sorted partition of a result set. The expression must return a single value. And it cannot be a window function.

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 9 above)
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: okay

File: custom/icds_reports/views.py (hunk 10 above)
Comment (challabeehyv) [reply custom/icds_reports/views.py]: changed it to GOVERNANCE_API_RECORDS_PAGINATION

File: custom/icds_reports/tasks.py (hunk 11 above)
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: you asked me to add this just like dashboard activity aggregation. I am using it for testing.Not sure the purpose

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 12 above)
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: makes sense.Should i change this?

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 8 above)
Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: I mean `FIRST_VALUE(anc_today) over w =1` is it same as `FIRST_VALUE(anm_mpw=1) over w`

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 12 above)
Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: yes please.

File: custom/icds_reports/tasks.py (hunk 11 above)
Comment (Rohit25negi) [reply custom/icds_reports/tasks.py]: I see.  well I think function `_agg_gov_vhnd_form` already servers that purpose. So you can remove this.

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 8 above)
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: yes. It should be the same

File: custom/icds_reports/const.py (hunk 14 above)
Comment (challabeehyv) [reply custom/icds_reports/const.py]: nope.removed

File: custom/icds_reports/tasks.py (hunk 13 above)
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: removed it

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 12 above)
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: done

File: custom/icds_reports/tasks.py (hunk 11 above)
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: yeah. removed

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
+from dateutil.relativedelta import relativedelta
```
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: is this file still needed?

File: custom/icds_reports/const.py [hunk 16]
```diff
@@ -111,6 +111,8 @@ class MapColors(object):
 AGG_THR_V2_TABLE = 'icds_dashboard_thr_v2'
 AGG_DASHBOARD_ACTIVITY = 'icds_dashboard_user_activity'
 AGG_ADOLESCENT_GIRLS_REGISTRATION_TABLE = 'icds_dashboard_adolescent_girls_registration'
+AGG_GOV_VHND_TABLE = 'icds_dashboard_gov_vhnd_forms'
```
Comment (Rohit25negi) [inline custom/icds_reports/const.py]: is this still needed after you shifted the vhnd to gov dashboard table?

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Left two comments. I think some cleanup is required.

File: custom/icds_reports/const.py (hunk 16 above)
Comment (challabeehyv) [reply custom/icds_reports/const.py]: removed

File: custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py (hunk 15 above)
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/gov_vhnd_forms.py]: removed

File: custom/icds_reports/tests/agg_tests/test_governance_apis.py [hunk 17]
```diff
@@ -84,9 +90,87 @@ def test_data_fetching_state_names_api(self):
...
+        """
+        test tp check if the first record is getting retrieved if there are multiple vhnds per month per awc
+        """
+        limit = GOVERNANCE_API_RECORDS_PAGINATION
+        query_filters = {'state_id': 'st1', 'awc_id__gt': 'a41', 'awc_launched': True}
+        order = ['awc_id']
+        data, count = get_vhnd_data(limit, 2017, 5, order, query_filters)
+        expected_row = {
+            "awc_id": "a49", "awc_code": "a49", "vhsnd_conducted": "yes",
+            "vhsnd_date": datetime.date(2017, 5, 2), "anm_present": "no", "asha_present": "yes",
```
Comment (Rohit25negi) [inline custom/icds_reports/tests/agg_tests/test_governance_apis.py]: I see in test fixture both forms have same vhnsd_date. Can you update second row in the fixture to later date to ensure its sorted by vhsnd_date.

File: custom/icds_reports/tests/agg_tests/test_governance_apis.py [hunk 18]
```diff
@@ -2,18 +2,24 @@
...
+from custom.icds_reports.reports.governance_apis import get_home_visit_data, get_state_names, get_vhnd_data,\
+    get_beneficiary_data
+
 from custom.icds_reports.tasks import _agg_governance_dashboard
 from datetime import date
 
+
 class GovernanceApiTest(TestCase):
 
+    def setUp(self):
```
Comment (Rohit25negi) [inline custom/icds_reports/tests/agg_tests/test_governance_apis.py]: good move!

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Overall looks good. lefts a few small clarity comments.

**Comment by Rohit25negi** [comment]
Also, Is it tested properly locally?

File: custom/icds_reports/reports/governance_apis.py [hunk 19]
```diff
@@ -31,6 +30,36 @@ def base_data(row_data):
...
+        anm_present=F('anm_mpw_present'), any_child_immunized=F('child_immu'),
+        anc_conducted=F('anc_today')).values(
+        'awc_id', 'awc_code', 'vhsnd_conducted', 'vhsnd_date', 'anm_present', 'asha_present',
+        'any_child_immunized', 'anc_conducted'
+    )
+
+    paginated_data = data[:length]
+
+    def get_value_or_data_not_entered(key, value):
+        if type(value) is bool or key == 'vhsnd_conducted':
```
[the remark this answers is not in the document]
Comment (challabeehyv) [reply custom/icds_reports/reports/governance_apis.py]: done

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_gov_dashboard.py [hunk 20]
```diff
@@ -143,6 +146,37 @@ def update_queries(self):
...
+            child_immu = ut.child_immu,
+            anc_today = ut.anc_today
+        FROM (
+            SELECT DISTINCT awc_id as awc_id,
+                FIRST_VALUE(vhsnd_date_past_month) over w as vhsnd_date_past_month,
+                FIRST_VALUE(anm_mpw=1) over w as anm_mpw_present,
+                FIRST_VALUE(asha_present=1) over w as asha_present,
+                FIRST_VALUE(child_immu=1) over w as child_immu,
+                FIRST_VALUE(anc_today=1) over w as anc_today,
+                %(start_date)s::DATE AS month
```
[the remark this answers is not in the document]
Comment (challabeehyv) [reply custom/icds_reports/utils/aggregation_helpers/distributed/agg_gov_dashboard.py]: no need since we are not doing the aggreagation

File: custom/icds_reports/tests/agg_tests/test_governance_apis.py (hunk 17 above)
Comment (challabeehyv) [reply custom/icds_reports/tests/agg_tests/test_governance_apis.py]: changed

File: custom/icds_reports/views.py [hunk 21]
```diff
@@ -2283,3 +2284,30 @@ class GovernanceStateListAPI(GovernanceAPIBaseView):
...
+    def get(self, request, *args, **kwargs):
+        last_awc_id, month, year, state_id = self.get_gov_api_params(request)
+        is_valid, error_message = self.validate_param(state_id, month, year)
+
+        if not is_valid:
+            return HttpResponse(error_message, status=400)
+
+        query_filters = {'awc_id__gt': last_awc_id, 'awc_launched': True}
+        order = ['awc_id']
+        if state_id != '':
```
[the remark this answers is not in the document]
Comment (challabeehyv) [reply custom/icds_reports/views.py]: to add it to the filter we need this check. But i see the base method is returning none for no value, changed it to none check

File: custom/icds_reports/reports/governance_apis.py [hunk 22]
```diff
@@ -31,6 +30,36 @@ def base_data(row_data):
...
 
 
+@icds_quickcache(['length', 'year', 'month', 'order', 'query_filters'], timeout=30 * 60)
+def get_vhnd_data(length, year, month, order, query_filters):
+    data = AggGovernanceDashboard.objects.filter(
+        month=date(year, month, 1),
+        **query_filters
+    ).order_by(*order).annotate(
+        vhsnd_conducted=F('vhsnd_date_past_month'), vhsnd_date=F('vhsnd_date_past_month'),
+        anm_present=F('anm_mpw_present'), any_child_immunized=F('child_immu'),
```
[the remark this answers is not in the document]
Comment (challabeehyv) [reply custom/icds_reports/reports/governance_apis.py]: refactoed it

File: custom/icds_reports/models/aggregate.py [hunk 23]
```diff
@@ -1586,6 +1586,12 @@ class AggGovernanceDashboard(models.Model, AggregateMixin):
     total_3_6_male_benefit_in_month = models.IntegerField(null=True)
     total_3_6_female_reg_in_month = models.IntegerField(null=True)
     total_3_6_male_reg_in_month = models.IntegerField(null=True)
+    vhsnd_date_past_month = models.DateField(null=True)
+    anm_mpw_present = models.NullBooleanField(null=True)
+    asha_present = models.NullBooleanField(null=True)
+    child_immu = models.NullBooleanField(null=True)
+    anc_today = models.NullBooleanField(null=True)
+    awc_id = models.TextField(primary_key=True)
```
[the remark this answers is not in the document]
Comment (challabeehyv) [reply custom/icds_reports/models/aggregate.py]: prob issue during merge conflict..removed it

File: custom/icds_reports/tests/agg_tests/test_governance_apis.py (hunk 18 above)
Comment (challabeehyv) [reply custom/icds_reports/tests/agg_tests/test_governance_apis.py]: Thanks

**Comment by challabeehyv** [comment]
> Also, Is it tested properly locally?

yes

**Review by Rohit25negi — APPROVED** [review]
Thanks for working on this. Looks good.
